contracts/external/cw-payroll/src/state.rs

```diff
@@ -0,0 +1,395 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, Decimal, Deps, OverflowError, Uint128};
+use cw_storage_plus::{Item, Map};
+
+use crate::ContractError;
+use cw_denom::{CheckedDenom, UncheckedDenom};
+
+use wynd_utils::Curve;
+
+#[cw_serde]
+pub struct UncheckedVestingParams {
+    pub recipient: String,
+    pub amount: Uint128,
+    pub denom: UncheckedDenom,
+    pub vesting_schedule: Curve,
+    pub title: Option<String>,
+    pub description: Option<String>,
+}
+
+#[cw_serde]
+pub struct CheckedVestingParams {
+    pub recipient: Addr,
+    pub amount: Uint128,
+    pub denom: CheckedDenom,
+    pub vesting_schedule: Curve,
+    pub title: Option<String>,
+    pub description: Option<String>,
+}
+
+impl UncheckedVestingParams {
+    pub fn into_checked(self, deps: Deps) -> Result<CheckedVestingParams, ContractError> {
+        // Check vesting schedule
+        self.assert_schedule_vests_amount()?;
+
+        // Check valid recipient address
+        let recipient = deps.api.addr_validate(&self.recipient)?;
+
+        // Check denom
+        let checked_denom = match self.denom {
+            UncheckedDenom::Native(denom) => UncheckedDenom::Native(denom).into_checked(deps)?,
+            UncheckedDenom::Cw20(addr) => UncheckedDenom::Cw20(addr).into_checked(deps)?,
+        };
+
+        // If title is included, validate title length (max 280 characters)
+        if let Some(ref title) = self.title {
+            if title.len() > 280 || title.is_empty() {
+                return Err(ContractError::InvalidTitle);
+            }
+        }
+
+        Ok(CheckedVestingParams {
+            recipient,
+            amount: self.amount,
+            denom: checked_denom,
+            vesting_schedule: self.vesting_schedule,
+            title: self.title,
+            description: self.description,
+        })
+    }
+
+    /// Asserts the vesting schedule decreases to 0 eventually, 2and is never more than the
+    /// amount being sent. If it doesn't match these conditions, returns an error.
+    pub fn assert_schedule_vests_amount(&self) -> Result<(), ContractError> {
+        self.vesting_schedule.validate_monotonic_decreasing()?;
+        let (low, high) = self.vesting_schedule.range();
+        if low != 0 {
+            Err(ContractError::NeverFullyVested)
+        } else if high > self.amount.u128() {
```

Comment:
how about requiring that `high == self.amount.u128()`? seems like if the curve would not vest enough that should also be an error